File: backend/src/main/java/com/example/SelfOrderingRestaurant/Service/AuthService.java
package com.example.SelfOrderingRestaurant.Service;

import com.example.SelfOrderingRestaurant.Dto.Request.UserRequestDTO.RegisterRequestDto;
import com.example.SelfOrderingRestaurant.Dto.Request.UserRequestDTO.*;
import com.example.SelfOrderingRestaurant.Dto.Response.UserResponseDTO.AuthResponseDto;
import com.example.SelfOrderingRestaurant.Entity.Customer;
import com.example.SelfOrderingRestaurant.Entity.Staff;
import com.example.SelfOrderingRestaurant.Entity.User;
import com.example.SelfOrderingRestaurant.Enum.UserStatus;
import com.example.SelfOrderingRestaurant.Enum.UserType;
import com.example.SelfOrderingRestaurant.Repository.CustomerRepository;
import com.example.SelfOrderingRestaurant.Repository.StaffRepository;
import com.example.SelfOrderingRestaurant.Repository.UserRepository;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.gson.GsonFactory;
import lombok.RequiredArgsConstructor;
import org.apache.tomcat.websocket.AuthenticationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.GeneralSecurityException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final StaffRepository staffRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenService jwtTokenService;
    private final AuthenticationManager authenticationManager;
    private final EmailService emailService;

    @Value("${google.client-id}")
    private String googleClientId;

    private final Logger log = LoggerFactory.getLogger(AuthService.class);

    @Transactional
    public AuthResponseDto registerStaff(RegisterRequestDto request, MultipartFile image) {
        // Check if username or email already exists
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new RuntimeException("Username already exists");
        }
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already exists");
        }

        // Create new user
        User user = new User();
        user.setUsername(request.getUsername());
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        user.setPhone(request.getPhone());
        user.setUserType(UserType.STAFF);
        user.setUserStatus(UserStatus.ACTIVE);
        user.setCreatedAt(LocalDateTime.now());

        User savedUser = userRepository.save(user);

        // Create staff profile
        Staff staff = new Staff();
        staff.setUser(savedUser);
        staff.setFullname(request.getFullname());
        staff.setHireDate(LocalDate.now());
        staff.setPosition("New Staff");
        staff.setSalary(BigDecimal.ZERO);
        staff.setStatus(UserStatus.ACTIVE);

        // Handle image if provided
        if (image != null && !image.isEmpty()) {
            String filePath = "src/main/resources/static/staff_faces/" + savedUser.getUserId() + ".jpg";
            try {
                Files.write(Paths.get(filePath), image.getBytes());
                staff.setFaceImagePath(filePath);
            } catch (IOException e) {
                throw new RuntimeException("Failed to save image: " + e.getMessage());
            }
        }

        Staff savedStaff = staffRepository.save(staff);

        // Generate tokens
        String accessToken = jwtTokenService.generateAccessToken(savedUser);
        String refreshToken = jwtTokenService.generateRefreshToken(savedUser);

        // Return response
        AuthResponseDto response = new AuthResponseDto();
        response.setAccessToken(accessToken);
        response.setRefreshToken(refreshToken);
        response.setUsername(savedUser.getUsername());
        response.setEmail(savedUser.getEmail());
        response.setUserType(savedUser.getUserType().name());
        response.setStaffId(savedStaff.getStaffId());

        return response;
    }

    public AuthResponseDto login(LoginRequestDto request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getLogin(),
                        request.getPassword()
                )
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);

        User user = userRepository.findByUsernameOrEmail(
                request.getLogin(),
                request.getLogin()
        ).orElseThrow(() -> new RuntimeException("User not found"));

        // Update last login
        user.setLastLogin(new Date());
        userRepository.save(user);

        // Generate tokens
        String accessToken = jwtTokenService.generateAccessToken(user);
        String refreshToken = jwtTokenService.generateRefreshToken(user);

        // Return response
        AuthResponseDto response = new AuthResponseDto();
        response.setAccessToken(accessToken);
        response.setRefreshToken(refreshToken);
        response.setUsername(user.getUsername());
        response.setEmail(user.getEmail());
        response.setUserType(user.getUserType().name());

        return response;
    }

    @Transactional
    public AuthResponseDto staffGoogleLogin(GoogleLoginRequestDto request) throws GeneralSecurityException, IOException {
        GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(
                new NetHttpTransport(),
                new GsonFactory()
        )
                .setAudience(Collections.singletonList(googleClientId))
                .setIssuer("https://accounts.google.com")
                .build();

        GoogleIdToken idToken = verifier.verify(request.getIdToken());

        if (idToken == null) {
            throw new RuntimeException("Token verification failed");
        }

        GoogleIdToken.Payload payload = idToken.getPayload();
        String email = payload.getEmail();
        String googleId = payload.getSubject();

        // Only accept email Gmail
        if (!email.endsWith("@gmail.com")) {
            throw new RuntimeException("Only Gmail accounts are allowed");
        }

        Optional<User> existingUser = userRepository.findByGoogleId(googleId);

        User user = existingUser.orElseGet(() -> {
            User newUser = new User();
            newUser.setGoogleId(googleId);
            newUser.setEmail(email);
            newUser.setUsername(email.split("@")[0]);
            newUser.setUserType(UserType.STAFF); // Default staff
            newUser.setUserStatus(UserStatus.ACTIVE);
            newUser.setCreatedAt(LocalDateTime.now());
            newUser.setLastLogin(new Date());

            return userRepository.save(newUser);
        });

        // Tạo token
        String accessToken = jwtTokenService.generateAccessToken(user);
        String refreshToken = jwtTokenService.generateRefreshToken(user);

        AuthResponseDto response = new AuthResponseDto();
        response.setAccessToken(accessToken);
        response.setRefreshToken(refreshToken);
        response.setUsername(user.getUsername());
        response.setEmail(user.getEmail());
        response.setUserType(user.getUserType().name());

        return response;
    }

    private void validateTokenPayload(GoogleIdToken.Payload payload) {
        if (payload == null) {
            throw new AuthenticationException("Invalid token payload");
        }

        // Additional checks
        if (!payload.getEmail().endsWith("@gmail.com")) {
            throw new AuthenticationException("Invalid email domain");
        }

        // Optional: Add more specific validation rules
        if (payload.getExpirationTimeSeconds() != null &&
                payload.getExpirationTimeSeconds() < System.currentTimeMillis() / 1000) {
            throw new AuthenticationException("Token has expired");
        }
    }

    public class AuthenticationException extends RuntimeException {
        public AuthenticationException(String message) {
            super(message);
        }
    }

    public void logout() {
        // Clear security context
        SecurityContextHolder.clearContext();
    }

    public AuthResponseDto refreshToken(RefreshTokenRequestDto request) {
        // Validate refresh token
        String username = jwtTokenService.extractUsername(request.getRefreshToken());
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Verify refresh token
        if (!jwtTokenService.validateRefreshToken(request.getRefreshToken(), user)) {
            throw new RuntimeException("Invalid refresh token");
        }

        // Generate new tokens
        String newAccessToken = jwtTokenService.generateAccessToken(user);
        String newRefreshToken = jwtTokenService.generateRefreshToken(user);

        // Return response
        AuthResponseDto response = new AuthResponseDto();
        response.setAccessToken(newAccessToken);
        response.setRefreshToken(newRefreshToken);
        response.setUsername(user.getUsername());
        response.setEmail(user.getEmail());
        response.setUserType(user.getUserType().name());

        return response;
    }

    @Transactional
    public void forgotPassword(ForgotPasswordRequestDto request) {
        // Validate email input
        if (request.getEmail() == null || request.getEmail().trim().isEmpty()) {
            throw new IllegalArgumentException("Email must not be empty");
        }

        if (request.getUsername() == null || request.getUsername().trim().isEmpty()) {
            throw new IllegalArgumentException("Username must not be empty");
        }

        // Optional: Add email format validation
        if (!isValidEmail(request.getEmail())) {
            throw new IllegalArgumentException("Invalid email format");
        }

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("User not found with email: " + request.getEmail()));

        // Generate reset token
        String otp = String.format("%06d", new Random().nextInt(999999));

        // Set reset token with explicit null checks
        user.setResetPasswordToken(otp);
        user.setResetPasswordExpiry(LocalDateTime.now().plusHours(10));

        try {
            userRepository.save(user);
            emailService.sendPasswordResetEmail(user.getEmail(), otp);
        } catch (Exception e) {
            // Log the actual exception
            throw new RuntimeException("Failed to process password reset", e);
        }
    }

    // Optional email validation method
    private boolean isValidEmail(String email) {
        String emailRegex = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$";
        return email != null && email.matches(emailRegex);
    }

    @Transactional
    public void resetPassword(ResetPasswordRequestDto request) {
        User user = userRepository.findByResetPasswordToken(request.getOtp())
                .orElseThrow(() -> new RuntimeException("Invalid reset token"));

        // Check token expiry
        if (user.getResetPasswordExpiry().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("OTP has expired");
        }

        // Update password
        user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        user.setResetPasswordToken(null);
        user.setResetPasswordExpiry(null);
        userRepository.save(user);
    }
}

File: frontend/src/services/authService.js
import api from "../axiosConfig";

class AuthService {
  // Login with username/email and password
  async login(login, password) {
    try {
      const response = await api.post("/auth/login", {
        login,
        password,
      });
      // Store tokens and user info in localStorage
      const { accessToken, refreshToken, username, userType } = response.data;
      if (accessToken) localStorage.setItem("accessToken", accessToken);
      if (refreshToken) localStorage.setItem("refreshToken", refreshToken);
      if (username) localStorage.setItem("username", username);
      if (userType) localStorage.setItem("userType", userType);
      return response.data;
    } catch (error) {
      const status = error.response?.status;
      let message = "Failed to login. Please try again.";
      if (status === 401) {
        message = "Invalid username or password.";
      } else if (status === 400) {
        message = error.response?.data?.message || "Invalid login request.";
      }
      throw new Error(message);
    }
  }

  // Login with Google
  async googleLogin(tokenId) {
    try {
      const response = await api.post("/auth/staff/google-login", {
        tokenId,
      });
      // Store tokens and user info in localStorage
      const { accessToken, refreshToken, username, userType } = response.data;
      if (accessToken) localStorage.setItem("accessToken", accessToken);
      if (refreshToken) localStorage.setItem("refreshToken", refreshToken);
      if (username) localStorage.setItem("username", username);
      if (userType) localStorage.setItem("userType", userType);
      return response.data;
    } catch (error) {
      const message =
        error.response?.data?.message ||
        "Failed to login with Google. Please try again.";
      throw new Error(message);
    }
  }

  // Logout
  async logout() {
    try {
      await api.post("/auth/logout");
      // Clear local storage
      this.clearLocalStorage();
      return { success: true };
    } catch (error) {
      console.error("Logout error:", error);
      // Clear storage even if API call fails
      this.clearLocalStorage();
      return {
        success: false,
        message: "Logout failed, but local data cleared.",
      };
    }
  }

  // Helper method to clear localStorage
  clearLocalStorage() {
    localStorage.removeItem("accessToken");
    localStorage.removeItem("refreshToken");
    localStorage.removeItem("username");
    localStorage.removeItem("userType");
  }

  // Request password reset
  async forgotPassword(login) {
    try {
      await api.post("/auth/forgot-password", { login });
      return { success: true, message: "Password reset link sent." };
    } catch (error) {
      const message =
        error.response?.data?.message ||
        "Failed to request password reset. Please try again.";
      throw new Error(message);
    }
  }

  // Refresh token
  async refreshToken(refreshToken) {
    try {
      const response = await api.post("/auth/refresh-token", {
        refreshToken,
      });
      // Update tokens in localStorage
      const { accessToken, refreshToken: newRefreshToken } = response.data;
      if (accessToken) localStorage.setItem("accessToken", accessToken);
      if (newRefreshToken)
        localStorage.setItem("refreshToken", newRefreshToken);
      return response.data;
    } catch (error) {
      const message =
        error.response?.data?.message || "Failed to refresh token.";
      throw new Error(message);
    }
  }

  // Get current user info from localStorage
  getCurrentUser() {
    return {
      username: localStorage.getItem("username"),
      userType: localStorage.getItem("userType"),
    };
  }

  // Check if user is authenticated
  isAuthenticated() {
    return !!localStorage.getItem("accessToken");
  }

  // Check if user has specific role
  hasRole(role) {
    const userType = localStorage.getItem("userType");
    return userType === role;
  }
}

export default new AuthService();
